### PR TITLE
Add converter method for ip adapters

### PR DESCRIPTION
**Pull request by fabiorigano**
# What does this PR do?

Fixes [5915 (comment)](https://github.com/huggingface/diffusers/pull/5915#discussion_r1413752931)

## Before submitting
- [x] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [x] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [ ] Was this discussed/approved via a GitHub issue or the [forum](https://discuss.huggingface.co/c/discussion-related-to-httpsgithubcomhuggingfacediffusers/63)? Please add a link to it if that's the case.
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@patrickvonplaten

### CODE REVIEW

**Comment by yiyixuxu** [comment]
can we maybe keep this function on the same file?

**Comment by fabiorigano** [comment]
> can we maybe keep this function on the same file?

done

I was wondering if it wouldn't be better to merge the two if blocks in `_load_ip_adapter_weights`, at line [713](https://github.com/huggingface/diffusers/blob/dfda680fea1fc186f79f6b2ec4b1fa559e37ae52/src/diffusers/loaders/unet.py#L713) and line [768](https://github.com/huggingface/diffusers/blob/dfda680fea1fc186f79f6b2ec4b1fa559e37ae52/src/diffusers/loaders/unet.py#L768)

File: src/diffusers/loaders/unet.py [hunk 1]
```diff
@@ -664,6 +664,44 @@ def delete_adapters(self, adapter_names: Union[List[str], str]):
             if hasattr(self, "peft_config"):
                 self.peft_config.pop(adapter_name, None)
 
+    def _convert_ip_adapter_to_diffusers(self, state_dict):
```
Comment (yiyixuxu) [inline src/diffusers/loaders/unet.py:667]:
```suggestion
    def _convert_ip_adapter_image_proj_to_diffusers(self, state_dict):
```

File: src/diffusers/loaders/unet.py [hunk 2]
```diff
@@ -789,36 +807,9 @@ def _load_ip_adapter_weights(self, state_dict):
...
-                elif "to_kv" in k:
-                    v_chunk = v.chunk(2, dim=0)
-                    new_sd[k.replace("to_kv", "to_k")] = v_chunk[0]
-                    new_sd[k.replace("to_kv", "to_v")] = v_chunk[1]
-                elif "to_out" in k:
-                    new_sd[k.replace("to_out", "to_out.0")] = v
-                else:
-                    new_sd[k] = v
-
+            new_sd = self._convert_ip_adapter_to_diffusers(image_proj_state_dict)
```
Comment (yiyixuxu) [inline src/diffusers/loaders/unet.py:810]: can we have this function return the `image_projection` module instead?

File: src/diffusers/loaders/unet.py (hunk 1 above)
Comment (sayakpaul) [reply src/diffusers/loaders/unet.py:667]: I wonder if we should move this method to https://github.com/huggingface/diffusers/blob/main/src/diffusers/loaders/ip_adapter.py. This way the UNet loader stays cleaner.

File: src/diffusers/loaders/unet.py [hunk 3]
```diff
@@ -737,18 +777,8 @@ def _load_ip_adapter_weights(self, state_dict):
...
-                    "image_embeds.weight": state_dict["image_proj"]["proj.weight"],
-                    "image_embeds.bias": state_dict["image_proj"]["proj.bias"],
-                    "norm.weight": state_dict["image_proj"]["norm.weight"],
-                    "norm.bias": state_dict["image_proj"]["norm.bias"],
-                }
-            )
-            image_projection.load_state_dict(image_proj_state_dict)
-            del image_proj_state_dict
+            new_sd = self._convert_ip_adapter_to_diffusers(image_proj_state_dict)
+            image_projection.load_state_dict(new_sd)
```
Comment (sayakpaul) [inline src/diffusers/loaders/unet.py:781]: Indeed better and cleaner!

File: src/diffusers/loaders/unet.py [hunk 4]
```diff
@@ -737,18 +777,8 @@ def _load_ip_adapter_weights(self, state_dict):
...
-                {
-                    "image_embeds.weight": state_dict["image_proj"]["proj.weight"],
-                    "image_embeds.bias": state_dict["image_proj"]["proj.bias"],
-                    "norm.weight": state_dict["image_proj"]["norm.weight"],
-                    "norm.bias": state_dict["image_proj"]["norm.bias"],
-                }
-            )
-            image_projection.load_state_dict(image_proj_state_dict)
-            del image_proj_state_dict
+            new_sd = self._convert_ip_adapter_to_diffusers(image_proj_state_dict)
```
Comment (sayakpaul) [inline src/diffusers/loaders/unet.py:780]:
```suggestion
            new_sd = self._convert_ip_adapter_image_proj_to_diffusers(image_proj_state_dict)
```

File: src/diffusers/loaders/unet.py [hunk 5]
```diff
@@ -759,20 +789,8 @@ def _load_ip_adapter_weights(self, state_dict):
...
-                    "ff.net.0.proj.bias": state_dict["image_proj"]["proj.0.bias"],
-                    "ff.net.2.weight": state_dict["image_proj"]["proj.2.weight"],
-                    "ff.net.2.bias": state_dict["image_proj"]["proj.2.bias"],
-                    "norm.weight": state_dict["image_proj"]["proj.3.weight"],
-                    "norm.bias": state_dict["image_proj"]["proj.3.bias"],
-                }
-            )
-            image_projection.load_state_dict(image_proj_state_dict)
-            del image_proj_state_dict
+            new_sd = self._convert_ip_adapter_to_diffusers(image_proj_state_dict)
```
Comment (sayakpaul) [inline src/diffusers/loaders/unet.py:792]:
```suggestion
            new_sd = self._convert_ip_adapter_image_proj_to_diffusers(image_proj_state_dict)
```

**Review by sayakpaul — COMMENTED** [review]
Really nice work!

@yiyixuxu can we also run the SLOW tests related to IP adapters with these changes before merging?

**Review by patrickvonplaten — APPROVED** [review]
Ok to merge for me once @yiyixuxu is happy with it

**Comment by fabiorigano** [comment]
@yiyixuxu @sayakpaul @patrickvonplaten thank you for your time and digging into the code :)

**Pull request merged**